JLLR-41: Feature/rest api

lakovicj opened the pull request.

lakovicj committed:
JLLR-41: To-do - Implement Restful API

lakovicj committed:
Renamed ToDo.php -> Todo.php

lakovicj committed:
JLLR-41: To-do - Implement Restful API

Fixed problem with implicit model binding during authorization

dejan-stokic commented on todo-backend/app/Models/User.php:
```diff
@@ -61,4 +61,8 @@ public function getJWTCustomClaims()
     }
+
+    public function todos() {
+        return $this->hasMany('App\Models\ToDo');
```
@lakovicj "Todo" umesto "ToDo".

dejan-stokic commented on todo-backend/app/Http/Controllers/API/TodoController.php:
```diff
@@ -0,0 +1,109 @@
+    public function store(Request $request)
+    {
+        $request->merge(array('priority' => strtolower($request->input('priority'))));
+        $request->validate([
```
@lakovicj Custom requests (FormRequest) su super stvar gde mozes definisati validaciju. To bi trebalo uci u praksu. Rasteretice kontroler.

lakovicj replied:
Ovo je izmesteno u FormRequests, svi se nalaze u App\Http\Requests.

dejan-stokic commented on todo-backend/app/Http/Controllers/API/TodoController.php:
```diff
@@ -0,0 +1,109 @@
+            'completed' => ['nullable', 'boolean']
+        ]);
+
+        $todo = new Todo();
```
@lakovicj Kontroleri trebaju biti "tanki", sva neka biznis logika treba ici u servise. Time granulisemo odgovornosti i imamo cistiji i reusable kod, lak se bilo kakve dalje promene.

Kreiraj app/Services/TodoService.php i prebaci tamo ovaj deo potreban za priperemu i kreiranje todo-a.

lakovicj replied:
Logika je izmestena u App\Services. Servisi su injektovani u konstruktoru kontrolera.

dejan-stokic commented on todo-backend/app/Http/Controllers/API/TodoController.php:
```diff
@@ -0,0 +1,109 @@
+     * @param  App\Models\Todo  $todo
+     * @return \Illuminate\Http\Response
+     */
+    public function update(Request $request, Todo $todo)
```
@lakovicj Isti komentari kao i za store(). Na kraju cemo imati samo par linija koda u odnosu na trenutno.

lakovicj committed:
Bug fixes

User was missing first_name and last_name fields. Also, register method is added in the AuthController. Request validations are isolated to FormRequests (StoreTodo, StoreUser, UpdateTodo). All the logic is moved out from controllers to Service layer (UserService, TodoService).

dejan-stokic commented on todo-backend/app/Http/Controllers/API/TodoController.php:
```diff
@@ -0,0 +1,97 @@
+    {
+        // priority to lower case
+        if ($request->priority != NULL) {
+            $request->merge(array('priority' => strtolower($request->input('priority'))));
```
@lakovicj Ovo isto moze u servis metodi da se uradi.

lakovicj committed:
Validation bugfix

Lower-casing request parameter 'priority' is moved out from controller's 'store' and 'update' method to FormRequest classes. createTodo method in TodoService now takes two arguments: first one is data, and second one is logged in user. Owner of Todo is now set in createTodo method.
